src/dbnode/storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -178,20 +301,515 @@ func (s *commitLogSource) ReadData(
+			return nil, fmt.Errorf(
+				"received bootstrap range that is not multiple of blockSize, blockSize: %d, start: %d, end: %d",
+				blockSize, currRange.End.Unix(), currRange.Start.Unix(),
+			)
+		}
+
+		// Reset this after we bootstrap each block to make sure its up to date.
+		if shardResult != nil {
+			allSeriesSoFar = shardResult.AllSeries()
+		}
+
+		for blockStart := currRange.Start.Truncate(blockSize); blockStart.Before(currRange.End); blockStart = blockStart.Add(blockSize) {
+			snapshotsForBlock := mostRecentCompleteSnapshotByBlockShard[xtime.ToUnixNano(blockStart)]
+			mostRecentCompleteSnapshotForShardBlock := snapshotsForBlock[shard]
+
+			if mostRecentCompleteSnapshotForShardBlock.CachedSnapshotTime.Equal(blockStart) ||
+				// Should never happen
+				mostRecentCompleteSnapshotForShardBlock.IsZero() {
+				// There is no snapshot file for this time, and even if there was, there would
+				// be no point in reading it. In this specific case its not an error scenario
+				// because the fact that snapshotTime == blockStart means we already accounted
+				// for the fact that this snapshot did not exist when we were deciding which
+				// commit logs to read.
+				s.log.Infof(
+					"no snapshots for shard: %d and blockStart: %d",
+					shard, blockStart.Unix())
+				continue
+			}
+
+			shardResult, err = s.bootstrapShardBlockSnapshot(
+				nsID, shard, blockStart, metadataOnly, shardResult, allSeriesSoFar, blockSize,
+				snapshotFiles, mostRecentCompleteSnapshotForShardBlock)
+			if err != nil {
+				return shardResult, err
+			}
+		}
+	}
+
+	if shardResult == nil {
+		shardResult = result.NewShardResult(0, s.opts.ResultOptions())
+	}
+	return shardResult, nil
+}
+
+func (s *commitLogSource) bootstrapShardBlockSnapshot(
+	nsID ident.ID,
+	shard uint32,
+	blockStart time.Time,
+	metadataOnly bool,
+	shardResult result.ShardResult,
+	allSeriesSoFar *result.Map,
+	blockSize time.Duration,
+	snapshotFiles fs.FileSetFilesSlice,
+	mostRecentCompleteSnapshot fs.FileSetFile,
+) (result.ShardResult, error) {
+	var (
+		bOpts      = s.opts.ResultOptions()
+		blOpts     = bOpts.DatabaseBlockOptions()
+		blocksPool = blOpts.DatabaseBlockPool()
+		bytesPool  = blOpts.BytesPool()
+		fsOpts     = s.opts.CommitLogOptions().FilesystemOptions()
+		idPool     = s.opts.CommitLogOptions().IdentifierPool()
+	)
+
+	// Bootstrap the snapshot file
+	reader, err := s.newReaderFn(bytesPool, fsOpts)
+	if err != nil {
+		return shardResult, err
+	}
+
+	err = reader.Open(fs.DataReaderOpenOptions{
+		Identifier: fs.FileSetFileIdentifier{
+			Namespace:   nsID,
+			BlockStart:  blockStart,
+			Shard:       shard,
+			VolumeIndex: mostRecentCompleteSnapshot.ID.VolumeIndex,
+		},
+		FileSetType: persist.FileSetSnapshotType,
+	})
+	if err != nil {
+		return shardResult, err
+	}
+
+	s.log.Infof(
+		"reading snapshot for shard: %d and blockStart: %d and volume: %d",
+		shard, blockStart.Unix(), mostRecentCompleteSnapshot.ID.VolumeIndex)
```

Comment:
Perhaps best to log this as string like we do elsewhere?  `blockStart.String()`

Reply:
done